aptos-move/framework/aptos-framework/sources/primary_fungible_store.move

```diff
@@ -137,7 +137,7 @@ module aptos_framework::primary_fungible_store {
     #[view]
     public fun is_balance_at_least<T: key>(account: address, metadata: Object<T>, amount: u64): bool {
         if (primary_store_exists(account, metadata)) {
-            fungible_asset::is_balance_at_least(primary_store(account, metadata), amount)
```

Comment:
This changed the optimization @igor-aptos wrote.

Reply:
Would be curious to see how much optimization it would change here? Seems like they should be the same?

Reply:
what does dispatchable balance means?

It sounds incompatible with concurrent balance (i.e. being able to do += , -= or >= on balance, without revealing balance itself). 

Does that mean that any dispatchable asset cannot be made concurrent?

For assets that don't have registered dispatchable functions, we cannot be revealing balance (in order to keep concurrency), so we would need to have dispatchable_fungible_asset::is_derived_balance_at_least, and use fungible_asset::is_balance_at_least if no dispatchable functions are registered. 

we should also confirm APT gas charging (as APT is guaranteed to be non-dispatchable) is still efficient.

Reply:
cc @georgemitenkov , @gelash for thoughts

Reply:
@runtian-zhou to clarify:

in fungible_asset.move, is_balance_at_least does:
  aggregator_v2::is_at_least(&balance_resource.balance, amount)
and balance does:
  aggregator_v2::read(&balance_resource.balance)

aggregator_v2::is_at_least is fully parallel. aggregator_v2::read is fully sequential.

Reply:
Oh I see. That's going to be crucial then.

Reply:
transfer is fully parallel , because it does:

```
is_balance_at_least(from, amount)
try_sub(from_balance, amount)
try_add(to_balance, amount)
```

for non-dispatchable tokens, it needs to stay that way. For dispatchable tokens, can it stay that way? or does it break the whole concurrency? Is there a move example code for dispatchable token, we can add to perf evaluation as well?

Reply:
Looking at the code there I think we might not want to change this. I reverted the relevant lines.

Reply:
hm, if you revert, does that mean that `is_balance_at_least` does a wrong thing for dispatchable assets?

Reply:
back to original question, what does "dispatchable_derived_balance" and "dispatchable_derived_supply" mean? is there an AIP with explanation of those functions?

Reply:
There wasn't. That was mostly added as a feature request from our partners.

Reply:
In this case, the dispatchable token is_greater_than function won't be parallelizable because we didn't register the hook, so we have to use the balance function to read out the u64 first.